src/server/search/search_family.cc

```diff
@@ -89,10 +90,87 @@ optional<SearchParams> ParseSearchParamsOrReply(CmdArgList args, ConnectionConte
         return nullopt;
       }
       i += 2;
+      continue;
+    }
+
+    // [PARAMS num(ignored) name(ignored) knn_vector]
+    if (ArgS(args, i) == "PARAMS") {
+      if (i + 3 >= args.size()) {
+        (*cntx)->SendError(kSyntaxErr);
+        return nullopt;
+      }
+      knn_vector = BytesToFtVector(ArgS(args, i + 3));
+      i += 3;
+      continue;
     }
   }
 
-  return SearchParams{limit_offset, limit_total};
+  return SearchParams{limit_offset, limit_total, move(knn_vector)};
+}
+
+void SendSerializedDoc(const SerializedSearchDoc& doc, ConnectionContext* cntx) {
+  (*cntx)->SendBulkString(doc.key);
+  (*cntx)->StartCollection(doc.values.size(), RedisReplyBuilder::MAP);
+  for (const auto& [k, v] : doc.values) {
+    (*cntx)->SendBulkString(k);
+    (*cntx)->SendBulkString(v);
+  }
+}
+
+void ReplyWithResults(const SearchParams& params, absl::Span<SearchResult> results,
+                      ConnectionContext* cntx) {
+  size_t total_count = 0;
+  for (const auto& shard_docs : results)
+    total_count += shard_docs.total_hits;
+
+  size_t response_count =
+      min(total_count - min(total_count, params.limit_offset), params.limit_total);
+
+  facade::SinkReplyBuilder::ReplyAggregator agg{cntx->reply_builder()};
+
+  (*cntx)->StartArray(response_count * 2 + 1);
+  (*cntx)->SendLong(total_count);
+
+  size_t sent = 0;
+  size_t to_skip = params.limit_offset;
+  for (const auto& shard_docs : results) {
+    for (const auto& serialized_doc : shard_docs.docs) {
+      // Scoring is not implemented yet, so we just cut them in the order they were retrieved
+      if (to_skip > 0) {
+        to_skip--;
+        continue;
+      }
+
+      if (sent++ >= response_count)
+        return;
+
+      SendSerializedDoc(serialized_doc, cntx);
+    }
+  }
+}
+
+void ReplyKnn(size_t knn_limit, const SearchParams& params, absl::Span<SearchResult> results,
+              ConnectionContext* cntx) {
+  vector<const SerializedSearchDoc*> docs;
+  for (const auto& shard_results : results) {
+    for (const auto& doc : shard_results.docs) {
+      docs.push_back(&doc);
+    }
+  }
+
+  partial_sort(docs.begin(), docs.end(),
```

Comment:
Haven't used it, but cppreference says `partial_sort` should be called with `(first, middle, last)`. This seems like a bug? Please add a test if it is.

Also, are the shard results sorted? The copy + partial sort is a bit redundant if they are.

Reply:
Good catch, wow. 

Yes, they are sorted. I know I can do it more efficiently for example by using a min heap and selecting how much is needed. Can add it later, either way the shard results are already capped.